File: 04월 08일 - 이분 탐색/과제/1477.py
import sys
input = sys.stdin.readline

"""
 [휴게소 세우기]

 M개의 휴게소를 짓고 난 후에 휴게소가 없는 구간의 최댓값의 최솟값은?
 -> 휴게소가 없는 구간의 최댓값이 k(임의의 수)일 때, M개의 휴게소를 지을 수 있는가?

 left: 휴게소가 없는 구간의 최솟값 -> 1
 right: 휴게소가 없는 구간의 최댓값 -> l - 1

 휴게소 구간의 최댓값이 주어졌을 때, 몇 개의 휴게소 짓는지 구하는 방법
 -> 처음 존재하는 휴게소 구간을 최대값으로 나누면 그 몫이 최대 구간이 최대값이 되도록 현재 구간 안에 설치해야 하는 휴게소 개수!
"""

# m개의 휴게소를 추가해서 휴게소 없는 구간이 dist 이하로 만들 수 있는지 확인하는 함수
def calc_rest(dist, n, location):
    count = 0
    for i in range(n-1):
        if location[i + 1] - location[i] > dist:
            count += (location[i + 1] - location[i] - 1) // dist
    
    return count

# 가능한 최댓값을 구하는 형식
def binary_search(n, m, l, location):
    left = 1
    right = l - 1
    while left <= right:
        mid = (left + right) // 2
        if calc_rest(mid, n, location) <= m:
            right = mid - 1
        else:
            left = mid + 1            
    return right + 1

# 입력
n, m, l = map(int, input().split())
location = list(map(int, input().split()))
# 도로의 시작과 끝을 추가하여, 도로 시작 ~ 첫 휴게소, 마지막 휴게소 ~ 도로 끝 구간도 체크할 수 있도록 한다.
location.append(0)
location.append(l)
location.sort() # 인접한 휴게소의 거리를 계산해야 하므로 정렬

print(binary_search(n+2, m, l, location))   # 양 끝값을 추가했으므로 n+2 대입

File: 04월 08일 - 이분 탐색/과제/1477.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

//휴게소 몇 개 설치하는지 구하는 함수
int cntRest(int dist, vector<int> &rest) {
    int cnt = 0;
    for (int i = 1; i < rest.size(); i++) {
        cnt += (rest[i] - rest[i - 1] - 1) / dist;
    }
    return cnt;
}

//휴게소가 없는 구간의 최댓값 중 최솟값 구하는 함수 -> lower bound
int lowerSearch(int left, int right, int m, vector<int> &rest) {
    while (left <= right) {
        int mid = (left + right) / 2; //최댓값 설정
        int cnt = cntRest(mid, rest); //몇 개 휴게소 설치하는지

        if (cnt > m) {
            left = mid + 1;
        } else {
            right = mid - 1;
        }
    }
    return left;
}

/**
 * [휴게소 세우기]
 *
 * M개의 휴게소를 짓고 난 후에 휴게소가 없는 구간의 최댓값의 최솟값은?
 * -> 휴게소가 없는 구간의 최댓값이 k(임의의 수)일 때, M개의 휴게소를 지을 수 있는가?
 *
 * left: 휴게소가 없는 구간의 최솟값 -> 1
 * right: 휴게소가 없는 구간의 최댓값 -> l - 1
 *
 * 휴게소 구간의 최댓값이 주어졌을 때, 몇 개의 휴게소 짓는지 구하는 방법
 * -> 처음 존재하는 휴게소 구간을 최대값으로 나누면 그 몫이 최대 구간이 최대값이 되도록 현재 구간 안에 설치해야 하는 휴게소 개수!
 */

int main() {
    int n, m, l;

    //입력
    cin >> n >> m >> l;
    vector<int> rest(n + 2, 0);
    for (int i = 1; i <= n; i++) {
        cin >> rest[i];
    }
    rest[n + 1] = l;

    //연산
    sort(rest.begin(), rest.end());

    //연산 & 출력
    cout << lowerSearch(1, l - 1, m, rest);
    return 0;
}